handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -845,9 +820,18 @@ public final SSLEngineResult unwrap(
             }
 
             if (len < SslUtils.SSL_RECORD_HEADER_LENGTH) {
-                return new SSLEngineResult(BUFFER_UNDERFLOW, getHandshakeStatus(), 0, 0);
-            }
-
+                return newResultMayFinishHandshake(BUFFER_UNDERFLOW, status, 0, 0);
+            }
+
+            // TODO(scott): limiting to a single packet can result in thrashing in the pipeline to decode application
+            // data. The javadocs elude to attempting to decode one complete SSL/TLS network packet [1], but I don't
+            // think this is meant to enforce a strict maximum of 1. The SslHandler's decode method already loops
+            // through the data to only give "full packets" to the SSLEngine implementations, we duplicate that
+            // logic here, and OpenSSL will also know how to delimit data. Can we remove this restriction which causes
+            // unnecessary overhead?
```

Comment:
@normanmaurer - great! I noticed this was thrashing a lot during my testing. let me break this out into another PR.